### ISSUE TITLE
Bug: Bot Purchase Frequency

### ISSUE THREAD

**Issue by Trus3683** [azerothcore/mod-ah-bot#109]
### Current Behaviour

Bot does not appear to be buying auctions very often.
Items are undercut from bot prices, usually very generously, but still nothing is purchased. 
Unknown - Is there a setting in the config that would adjust for this? 

### Expected Behaviour

Bot should purchase auctions, or there should be a setting to determine how many auctions are purchased.

### Steps to reproduce the problem

1. Post auction.
2. No purchase. 

### Extra Notes

N/A

### AC rev. hash/commit

N/A

### Operating system

Windows 11

**Comment by kewinrausch** [comment]
You have to setup the `buyerbidsperinterval`  value present into the `mod_auctionhousebot` table for the desired auction house. The default is set to buy one element per minute less or more.

**Comment by Trus3683** [comment]
> You have to setup the `buyerbidsperinterval` value present into the `mod_auctionhousebot` table for the desired auction house. The default is set to buy one element per minute less or more.

Hello, I changed the value to 10 and it doesn't seem to have done anything - Is there anything else? Before, the bot was buying things like crazy and I didn't modify any data elements.

**Comment by danielcoh92** [comment]
I have found out that the buyer function only works if you give a character GUID for "AuctionHouseBot.GUID" parameter
Otherwise, nothing happens.
I tried all use-cases I could think of and this is the only way to make the buyer function work for me.

**Comment by Trus3683** [comment]
> I would say you have to enable the `TRACE_BUYER` setting it to 1 and monitor what is happening, or eventually setup the `DEBUG_BUYER` and see what the system is telling you. Can you also share the config file and the table content so I can take a look on the configuration situation?
[mod_ahbot.txt](https://github.com/user-attachments/files/17267086/mod_ahbot.txt)
Config file provided. How would I provide the table content?

**Comment by kewinrausch** [comment]
It exists a block in the Buy routine that does not allow the bots to buy stuff from another bot. Maybe this is what you are experiencing. To avoid generating noise on the market right now bots offers lot of auctions to players and consume only players auctions, since they exists only to supports players on low population servers.

To provide the table content, you can simply copy and paste the `select * from mod_auctionhousebot;` SQL command in a txt file and post it here.

**Comment by Trus3683** [comment]
[ahbot.txt](https://github.com/user-attachments/files/17268431/ahbot.txt)
Does this work? Thanks!

**Comment by kewinrausch** [comment]
If you see the content of the data you provided to me you will see that `buyerbidsperinterval` is set to 10, which means that around every minute (`buyerbiddinginterval`) the bot will try to bid for 10 auctions (bid does not mean buy, it just offers something and when the auction expire the bot will win if it's the last bidder).

Of this 10 attempts, the bots will NOT bid for auctions of other bots, or if the given offers are outside its bidding limits (the bot will not buy that raptor flesh for 5000 golds). There are lot of variables to take into account. For example it depends if it has to look for buy or sell standard prices, and if these standard prices are specified (some objects present into the market could have no prices specified into the database, like for some enchanting reagents).

To sum up: it will BID (sometime BUY) items with a standard price specified offered within a reasonable delta of that price, that do not belong to other bots.

**Comment by Trus3683** [comment]
> If you see the content of the data you provided to me you will see that `buyerbidsperinterval` is set to 10, which means that around every minute (`buyerbiddinginterval`) the bot will try to bid for 10 auctions (bid does not mean buy, it just offers something and when the auction expire the bot will win if it's the last bidder).
> 
> Of this 10 attempts, the bots will NOT bid for auctions of other bots, or if the given offers are outside its bidding limits (the bot will not buy that raptor flesh for 5000 golds). There are lot of variables to take into account. For example it depends if it has to look for buy or sell standard prices, and if these standard prices are specified (some objects present into the market could have no prices specified into the database, like for some enchanting reagents).
> 
> To sum up: it will BID (sometime BUY) items with a standard price specified offered within a reasonable delta of that price, that do not belong to other bots.

What do I do to increase the frequency that the bot will BUY items? As mentioned before, it is buying significantly less frequently than it has in the past.

**Comment by kewinrausch** [comment]
The buy mechanism will buy if the price chosen for the bid is high enough.
Try to put larger values for the `minbidpriceXXX` and `maxbidpriceXXX` fields in the database; this should make the bot more generous and allocate larger sum to bet. If the sum is large enough it will perform a straight buyout instead of a bid. That and increasing `buyerbidsperinterval` should do the job.

**Comment by Bits-per-post** [comment]
I can confirm altering the buyerbidsperinterval value in the database previously increased bot bids and purchase frequency but in newer builds altering this value no longer has the desired effect.

While I could spend an eternity going through the code for the latest commits (professions items etc) instead I simply opted to edit the source code and rebuild.

navigate to modules / mod-ah-bot / src and edit the AuctionHouseBot.cpp file

find the start of buyer loop in the code

" //
  // Perform the operation for a maximum amount of bids attempts configured
  //

    for (uint32 count = 1; count <= config->GetBidsPerInterval(); ++count)
    {
"
and edit  "count <= config->GetBidsPerInterval() "   to the number of bids you want.   for me as a solo player i changed to 1000 and it simply bids/busy all my items instantly without breaking too much.   not recommended if you are running a server, troubleshoot it properly and find where in the code the buyerbidsperinterval function is being skipped over and correct it.   but if you just want to see immediate effect, changing this value simply works without breaking too much.

**Comment by Tapap84** [comment]
> I have found out that the buyer function only works if you give a character GUID for "AuctionHouseBot.GUID" parameter Otherwise, nothing happens. I tried all use-cases I could think of and this is the only way to make the buyer function work for me.

Im having the same issue. How do I check the guid number? My build is straight on debian using mysql

**Comment by wenzhizhang** [comment]
I met the same issue. For my case in the server log below message can be found:
`Auction id: 10644 Possible entry to buy/bid from AH pool is invalid, this should not happen, moving on next auciton`

I checked the source code and found the logic below:
`QueryResult ahContentQueryResult = CharacterDatabase.Query("SELECT id FROM auctionhouse WHERE houseid={} AND itemowner<>{} AND buyguid<>{}", config->GetAHID(), _id, _id);`

Seems there was dirty data in this table `auctionhouse  `in database `acore_characters`, so I just deleted all rows in this table then the ah bot started to buy things. 

For your reference.
